### ISSUE TITLE
Rule Change: Allow generator `return` statements inside `finally` blocks in `no-unsafe-finally`

### ISSUE THREAD

**Issue by sethamus** [eslint/eslint#20332]
### What rule do you want to change?

`no-unsafe-finally`

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
function* gen() {
  try {
    yield 1;
  } finally {
    return "cleanup";
  }
}
```

### What does the rule currently do for this code?

`no-unsafe-finally` reports an error:
```
Unsafe usage of ReturnStatement.
```

The rule treats any `return` inside a `finally` block as unsafe because it overrides the `try` block’s completion value. However, in generator functions, this pattern is documented and intentional. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return), calling `return()` on a generator causes execution to proceed to the `finally` block if the generator is suspended within a `try`, and a `return` inside that `finally` becomes the value returned from `generator.return()`.

### What will the rule do after it's changed?

The rule should allow `return` statements inside `finally` blocks when the surrounding function is a generator function.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

https://github.com/eslint/eslint/issues/20313

**Comment by nzakas** [comment]
This is interesting and a bit nuanced. I'm not sure I'm ready to say that such a pattern is actually safe because it's still within the context of a `try-finally`, which means if there's a `return` in `try` then the return in `finally` takes precedence. We could potentially add an option to allow this pattern, but I'd like some more opinions.

@eslint/eslint-team

**Comment by Josh-Cena** [comment]
Hi, I'm the author of the MDN doc you are quoting. What I wrote merely reflects what the language _allows_ you to do; it doesn't really say if you actually _should_ do that, and I don't know if this is good—do you have a valid use case IRL?

**Comment by TomStrepsil** [comment]
> Hi, I'm the author of the MDN doc you are quoting. What I wrote merely reflects what the language _allows_ you to do; it doesn't really say if you actually _should_ do that, and I don't know if this is good—do you have a valid use case IRL?

I assumed it was to allow getting the "remainder" when cancelling generation, e.g. something like:
```ts
function* process(inputString: string): Generator<string> {
  let position;
  try {
    for (position = 0; position < inputString.length; position++) {
      yield doSomeOnDemandThing(inputString[position]);
    }
  } finally {
    return inputString.slice(position);
  }
}
```
...allows a consumer to generate the output of `doSomeOnDemandThing` till some limit / trigger, then get access to "what's left" using [`return()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return).

I accept that `yield` within the `finally` achieves _similar_.  I guess `return` gives the generator the option to _reject_ the value supplied to the function, rather than have it yield afterwards.  It also ensuring the generator is `{ done: true }` after calling `return()`, rather than permitting the next call to `.next()` to return something.

I appreciate, that's not a use-case demonstrated above.

**Comment by nzakas** [comment]
ping @eslint/eslint-team

**Comment by mdjermanovic** [comment]
This seems like an edge case for which we are also not sure if it is a safe and common practice, so I think it's fine to just recommend using eslint-disable comments as already noted in the docs (https://github.com/eslint/eslint/pull/20330).

**Comment by Josh-Cena** [comment]
Although uncommon, I don't see anything particularly bad about this, unlike normal return in finally. Normally, if you return in finally, then the original completion gets ignored, and linters are supposed to catch unused/masked things. But in this case, the _unmasked_ return value is the useless one, since the caller already knows what that's gonna be; the only interesting case is where a value _does_ get returned inside `finally`, so it seems there's nothing that goes against the spirit of the rule here.

**Comment by kirkwaiblinger** [comment]
Surely this swallows exceptions just as dangerously as a normal return-in-finally-block?

```js
function * throwingGenerator() {
  yield 1;
  throw new Error('Important error occurred in throwingGenerator');
  yield 2; // unreachable
}

function * wrappingGenerator() {
  try {
    yield 'start';
    yield * throwingGenerator(); // or substitute any throwing operation
    yield 'end'; 
  } finally{
    console.log('Cleanup in finally block of wrappingGenerator');
    // return 'cleanup'; <-- uncommenting this line swallows the exception, preventing it from propagating
  }
}


for (const value of wrappingGenerator()) {
    console.log('value:', value);
}
console.log('exited generator normally');
```

If we want to give ourselves a headache, we could also think about what happens when calling `generator.throw()` and `generator.return()`... (bottom line is the same; the `return` statement in the generator discards the result which would be specified by `generator.throw()`/`.return()`)

The given use case is intriguing:
```js
function* process(inputString) {
  let position;
  try {
    for (position = 0; position < inputString.length; position++) {
      yield doSomeOnDemandThing(inputString[position]);
    }
  } finally {
    return inputString.slice(position);
  }
}
```
Is a method that would silently also discard any exception thrown by `doSomeOnDemandThing()` the best way to cancel a generator, though? Another idea that only responds to intentional, explicit cancellations would be:
```js
/**
 * To exit early, call `generator.next(true)` (or supply any truthy value).
 */
function* process(inputString) {
  let position;
  for (position = 0; position < inputString.length; position++) {
    const exitEarly = yield doSomeOnDemandThing(inputString[position]);
    if (exitEarly) {
      break;
    }
  }
  return inputString.slice(position);
}
```

---

I get that sometimes you _want_ to swallow exceptions... but that is spelled `catch`, not `finally`, just as with a normal function, no?

**Comment by kirkwaiblinger** [comment]
FWIW - I don't even think that doing cancellations as part of the generator protocol methods is even a good idea, since the language doesn't support it syntactically (meaning you _have_ to iterate the generator with `gen.next()` and the like). Consider instead an explicit protocol like so:

```js
class CancellationError extends Error {
  constructor() {
    super('Operation was cancelled');
    this.name = 'CancellationError';
  }
}

class CancellationToken {
  isCancelled = false;

  /* 
   * Could be extended to take a parameter that adds data into the error;
   * this would have the same utility as the return value in TomStrepsil's example.
   */
  throwIfCancelled() {
    if (this.isCancelled) {
      throw new CancellationError();
    }
  }

  cancel() {
    this.isCancelled = true;
  }
}

function * cancellableGenerator(cancellationToken = undefined) {
  cancellationToken?.throwIfCancelled();
  yield someExpensiveThing();
  cancellationToken?.throwIfCancelled();
  yield someOtherExpensiveThing();
}

try {
  // can use the generator in an ordinary for...or loop without modification!
  const token = new CancellationToken();
  for (const item of cancellableGenerator(token)) {
    console.log(item);
    if (someCondition()) {
      token.cancel();
    }
  }
} catch (e) {
  if (!(e instanceof CancellationError)) {
    throw e;
  }
  // handle cancellation specifically
}
```

**Comment by TomStrepsil** [comment]
@kirkwaiblinger - I would concur, having raised the original documentation issue (https://github.com/eslint/eslint/issues/20313), I frankly had not considered the error-swallowing nature of `return` within `finally`, despite being [an example in the documentation](https://eslint.org/docs/latest/rules/no-unsafe-finally#:~:text=//%20We%20expect%20this%20function%20to%20throw%20an%20error%2C%20then%20return).

It feels like a dangerous thing to permit.

On the cancellation token concept, I think an [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) might be a good choice.

e.g. consider this example - a generator with `finally` that yields when `return()` called on the generator, prompted by an `AbortSignal`:
```js
const letters = "abcdefghijklmnopqrstuvwxyz";

function* generate() {
  let i;
  try {
    for (i = 0; i < letters.length; i++) {
      yield letters[i].toUpperCase();
    }
  } finally {
    yield letters.slice(i);
  }
}

function* abortable(signal) {
  const generator = generate();
  for (const letter of generator) {
    if (signal.aborted) {
      yield generator.return().value;
      break;
    }
    yield letter;
  }
}

const abortController = new AbortController();
for (const letter of abortable(abortController.signal)) {
  if (letter === "N") {
    abortController.abort();
  }
  process.stdout.write(letter);
}
```
This outputs `ABCDEFGHIJKLMNopqrstuvwxyz`, via uppercase letter-by-letter up till "N", then sends an abort signal to the abortable generator wrapper, to yield the output of `return()` on an underlying generator.

**Comment by mdjermanovic** [comment]
Since this change request has been open for a month but hasn't received a +1 from any team member, and by the discussion it doesn't seem we should consider this pattern safe, I am closing it. A note about using `eslint-disable` comments for this use case has already been added to the docs: https://eslint.org/docs/next/rules/no-unsafe-finally#when-not-to-use-it.
